Upgrade to serde 0.9

jrasky opened the pull request:
Rework things, refactor and fix everything to support serde 0.9.

jrasky committed:
WIP Do some work to update ser and error

jrasky committed:
WIP It compiles again, but fails tests

jrasky committed:
Refactor things out of ser

Pull SeqSerializer and MapVariantSerializer out of ser and into their own modules

jrasky committed:
Remove some dead code

jrasky committed:
Refactor capabilites out of eachother

MapSerializer is now separate from SeqSerializer

jrasky committed:
Fix a bug with map length

jrasky committed:
refactor(de): split up de a bit

Move each struct into its own module rather than having them all be in the same one

jrasky committed:
fix(de, ser): Resolve remaining bugs

All tests now pass

jrasky committed:
cleanup(): Delint everything

jrasky committed:
feature(ser): avoid allocation where possible

jrasky committed:
fix(ser): Check result instead of ignoring it

jrasky committed:
Bump version number

jrasky committed:
merge(): Merge in release changes from master

jrasky committed:
fix(lib): Remove duplicate definition

jrasky committed:
doc(): Add clobbered docstrings

jrasky committed:
refactor(de): Remove helper for input function

Access the field appropriately at the call site in all cases

jrasky committed:
doc(): Add license headers and descriptions

jrasky committed:
doc(): Update readme and dependency version policy

Use tilde matching to better specify versions for everything

jrasky committed:
fix(lib): Remove old feature gate

jrasky committed:
delint(): run rustfmt

jrasky commented:
@astraw take a look

dtolnay commented on Cargo.toml:
```diff
@@ -11,11 +10,11 @@ keywords = ["messagepack", "binary", "encoding", "serialize"]
-[build-dependencies]
-serde_codegen = { version = "0.9" }
+[dev-dependencies]
+serde_derive = { version = "~0.9.8", features = ["unstable"] }
```
What unstable feature do you need from serde_derive? I think this cfg actually doesn't do anything in 0.9.8.

jrasky committed:
delint(test): Remove unused dependency feature

"unstable" doesn't do anything in serde_derive

jrasky committed:
test(): Add failing test for Option

jrasky committed:
wip(de): Work on deserialize_option

Do some thinking about how to handle Option types in Deserializer

jrasky committed:
fix(de): Fix Option deserialization

Add some state to track the type encountered, so that visit_some works correctly

dtolnay commented:
You can skip 0.9 and go directly to Serde 1.0! https://github.com/serde-rs/serde/releases/tag/v1.0.0

jrasky committed:
wip(): Update to serde 1.0

Many things still broken

jrasky committed:
feat(): Upgrade to serde 1.0

TODO refactor and cleanup

dtolnay commented:
The `from_iter` and `from_bytes` signatures should be:

```rust
pub fn from_iter<I, V>(iter: I) -> Result<V>
    where I: Iterator<Item = u8>,
          V: DeserializeOwned;

pub fn from_bytes<'a, V>(bytes: &'a [u8]) -> Result<V>
    where V: Deserialize<'a>;
```

In `from_iter` it requires that V does not try to borrow any data from the input because iterators don't support that. In `from_bytes` it says V is allowed to borrow from the input bytes of lifetime `'a`.

jrasky commented:
I need to actually pour over the lifetime documentation for serde. I hacked at it until it compiled, but I'm confident that it's not in good shape. Thanks for pointing that out!

jrasky committed:
feat(): Bring back our own Error type

jrasky committed:
wip(): Work on zero-copy support

jrasky committed:
wip(): Implement zero-copy

TODO resolve remaining bugs

jrasky committed:
feat(): Resolve issues with zero-copy

jrasky committed:
fix(test): DeserializeOwned is easier to read

jrasky committed:
doc(): Add documentation for Read

Also run rustfmt

jrasky committed:
feat(): Turn on std by default

In order to match with the default for serde

jrasky commented:
Okay, I think things are in decent shape. I'll merge this in and push to crates.io on monday.

astraw committed:
feat(): runs on stable rust

astraw commented:
@jrasky it is working for me with serde 1.0 on nightly (tests run and my code runs both with and without std). With my PR #6, this is true also for stable rust. Looking forward to the release.

astraw committed:
feat(): combine all nightly features into "collections" feature

jrasky committed:
Merge pull request #6 from astraw/stable-serde-1

feat(): runs on stable rust

jrasky committed:
fix(ser, de): Compiles with all all features

Also bump the listed serde version

jrasky committed:
doc(README): Add a note about feature options
